File: web/src/features/workout/api/workoutApi.ts
// =====================================================================
// KineticOs — Cliente REST del módulo workout (/api/v1/workouts).
// Paridad con mobile/lib/features/workout/infrastructure/workout_api.dart.
// =====================================================================
import { apiClient } from '@/core/api/client';
import type {
  Exercise,
  ExerciseFilter,
  Muscle,
  SessionExercise,
  Workout,
  WorkoutDraft,
  WorkoutSession,
} from '../domain/workoutTypes';
import { workoutDraftToPayload } from '../domain/workoutTypes';

export const workoutApi = {
  listMuscles: () => apiClient.get<Muscle[]>('/workouts/muscles').then((r) => r.data),

  listExercises: (filter: ExerciseFilter) =>
    apiClient
      .get<Exercise[]>('/workouts/exercises', {
        params: {
          category: filter.category,
          difficulty: filter.difficulty,
          muscleId: filter.muscleId,
          search: filter.search || undefined,
        },
      })
      .then((r) => r.data),

  getExercise: (id: number) => apiClient.get<Exercise>(`/workouts/exercises/${id}`).then((r) => r.data),

  listWorkouts: (templates: boolean) =>
    apiClient.get<Workout[]>('/workouts', { params: { templates } }).then((r) => r.data),

  getWorkout: (id: number) => apiClient.get<Workout>(`/workouts/${id}`).then((r) => r.data),

  createWorkout: (draft: WorkoutDraft) =>
    apiClient.post<Workout>('/workouts', workoutDraftToPayload(draft)).then((r) => r.data),

  generateWorkout: (goal: string) =>
    // La IA local (Ollama, CPU) puede tardar bastante más que una llamada normal.
    apiClient.post<Workout>('/workouts/generate', { goal }, { timeout: 180000 }).then((r) => r.data),

  updateWorkout: (id: number, draft: WorkoutDraft) =>
    apiClient.put<Workout>(`/workouts/${id}`, workoutDraftToPayload(draft)).then((r) => r.data),

  deleteWorkout: (id: number) => apiClient.delete(`/workouts/${id}`),

  listSessions: () => apiClient.get<WorkoutSession[]>('/workouts/sessions').then((r) => r.data),

  getSession: (id: number) => apiClient.get<WorkoutSession>(`/workouts/sessions/${id}`).then((r) => r.data),

  startSession: (body: { workoutId?: number | null; workoutDayId?: number | null }) =>
    apiClient.post<WorkoutSession>('/workouts/sessions', body).then((r) => r.data),

  completeSession: (id: number, body: { durationSeconds?: number | null; notes?: string | null }) =>
    apiClient.put<WorkoutSession>(`/workouts/sessions/${id}/complete`, body).then((r) => r.data),

  abortSession: (id: number, body: { notes?: string | null }) =>
    apiClient.put<WorkoutSession>(`/workouts/sessions/${id}/abort`, body).then((r) => r.data),

  logSessionExercise: (
    sessionId: number,
    body: {
      workoutExerciseId?: number | null;
      exerciseId: number;
      setsDone?: number | null;
      weightKg?: number | null;
      reps?: number | null;
      rpe?: number | null;
      durationSeconds?: number | null;
      distanceMeters?: number | null;
      notes?: string | null;
    },
  ) => apiClient.post<SessionExercise>(`/workouts/sessions/${sessionId}/exercises`, body).then((r) => r.data),
};

File: mobile/lib/features/workout/infrastructure/workout_api.dart
// =====================================================================
// KineticOs — Cliente REST del módulo workout (/api/v1/workouts).
// =====================================================================
import 'package:dio/dio.dart';

import '../domain/exercise.dart';
import '../domain/exercise_filter.dart';
import '../domain/muscle.dart';
import '../domain/workout.dart';
import '../domain/workout_draft.dart';
import '../domain/workout_session.dart';

class WorkoutApi {
  final Dio _dio;

  WorkoutApi(this._dio);

  // ------------------------- Catálogo -------------------------

  Future<List<Muscle>> listMuscles() async {
    final response = await _dio.get('/workouts/muscles');
    return (response.data as List<dynamic>).map((e) => Muscle.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<List<Exercise>> listExercises(ExerciseFilter filter) async {
    final response = await _dio.get('/workouts/exercises', queryParameters: filter.toQueryParams());
    return (response.data as List<dynamic>).map((e) => Exercise.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Exercise> getExercise(int id) async {
    final response = await _dio.get('/workouts/exercises/$id');
    return Exercise.fromJson(response.data as Map<String, dynamic>);
  }

  // ------------------------- Rutinas -------------------------

  Future<List<Workout>> listWorkouts({required bool templates}) async {
    final response = await _dio.get('/workouts', queryParameters: {'templates': templates});
    return (response.data as List<dynamic>).map((e) => Workout.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<Workout> getWorkout(int id) async {
    final response = await _dio.get('/workouts/$id');
    return Workout.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Workout> createWorkout(WorkoutDraft draft) async {
    final response = await _dio.post('/workouts', data: draft.toJson());
    return Workout.fromJson(response.data as Map<String, dynamic>);
  }

  Future<Workout> updateWorkout(int id, WorkoutDraft draft) async {
    final response = await _dio.put('/workouts/$id', data: draft.toJson());
    return Workout.fromJson(response.data as Map<String, dynamic>);
  }

  Future<void> deleteWorkout(int id) => _dio.delete('/workouts/$id');

  /// Genera una rutina con IA a partir de un pedido en lenguaje natural. Puede tardar
  /// hasta un par de minutos (proveedor local por CPU) — timeout largo a propósito.
  Future<Workout> generateWorkout(String goal) async {
    final response = await _dio.post(
      '/workouts/generate',
      data: {'goal': goal},
      options: Options(sendTimeout: const Duration(seconds: 10), receiveTimeout: const Duration(seconds: 180)),
    );
    return Workout.fromJson(response.data as Map<String, dynamic>);
  }

  // ------------------------- Sesiones -------------------------

  Future<List<WorkoutSession>> listSessions() async {
    final response = await _dio.get('/workouts/sessions');
    return (response.data as List<dynamic>)
        .map((e) => WorkoutSession.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<WorkoutSession> getSession(int id) async {
    final response = await _dio.get('/workouts/sessions/$id');
    return WorkoutSession.fromJson(response.data as Map<String, dynamic>);
  }

  Future<WorkoutSession> startSession({int? workoutId, int? workoutDayId}) async {
    final response = await _dio
        .post('/workouts/sessions', data: {'workoutId': workoutId, 'workoutDayId': workoutDayId});
    return WorkoutSession.fromJson(response.data as Map<String, dynamic>);
  }

  Future<WorkoutSession> completeSession(int id, {int? durationSeconds, String? notes}) async {
    final response = await _dio.put('/workouts/sessions/$id/complete',
        data: {'durationSeconds': durationSeconds, 'notes': notes});
    return WorkoutSession.fromJson(response.data as Map<String, dynamic>);
  }

  Future<WorkoutSession> abortSession(int id, {String? notes}) async {
    final response = await _dio.put('/workouts/sessions/$id/abort', data: {'notes': notes});
    return WorkoutSession.fromJson(response.data as Map<String, dynamic>);
  }

  Future<SessionExercise> logSessionExercise(
    int sessionId, {
    int? workoutExerciseId,
    required int exerciseId,
    int? setsDone,
    double? weightKg,
    int? reps,
    int? rpe,
    int? durationSeconds,
    int? distanceMeters,
    String? notes,
  }) async {
    final response = await _dio.post('/workouts/sessions/$sessionId/exercises', data: {
      'workoutExerciseId': workoutExerciseId,
      'exerciseId': exerciseId,
      'setsDone': setsDone,
      'weightKg': weightKg,
      'reps': reps,
      'rpe': rpe,
      'durationSeconds': durationSeconds,
      'distanceMeters': distanceMeters,
      'notes': notes,
    });
    return SessionExercise.fromJson(response.data as Map<String, dynamic>);
  }
}
